firedrake/bcs.py

```diff
@@ -297,3 +396,93 @@ def homogenize(bc):
+    :arg method: see :class:`.DirichletBC` (optional)
+    :arg sub_space_index: the sub_space index for the function space
+        on which the equation boundary condition is applied
+    '''
+
+    def __init__(self, eq, u, sub_domain, bcs=None, J=None, Jp=None, method="topological", V=None, is_linear=False):
+
+        if V is None:
+            V = eq.lhs.arguments()[0].function_space()
+        super().__init__(V, sub_domain, method=method)
+        # u is always the total solution just as in "solve(F==0, u, ...)"
+        self.u = u
+        # This nested structure will enable recursive application of boundary conditions.
+        #
+        # def _assemble(..., bcs, ...)
+        #     ...
+        #     for bc in bcs:
+        #         # boundary conditions for boundary conditions for boun...
+        #         ... _assemble(bc.f, bc.bcs, ...)
+        #     ...
+        self.bcs = bcs or []
+
+        self.Jp_eq_J = Jp is None
+        self.is_linear = is_linear
+        # linear
+        if isinstance(eq.lhs, ufl.Form) and isinstance(eq.rhs, ufl.Form):
+            self.J = eq.lhs
+            self.Jp = Jp or self.J
+            if eq.rhs == 0:
+                self.F = ufl_expr.action(self.J, self.u)
+            else:
+                if not isinstance(eq.rhs, (ufl.Form, slate.slate.TensorBase)):
+                    raise TypeError("Provided BC RHS is a '%s', not a Form or Slate Tensor" % type(eq.rhs).__name__)
+                if len(eq.rhs.arguments()) != 1:
+                    raise ValueError("Provided BC RHS is not a linear form")
+                self.F = ufl_expr.action(self.J, self.u) - eq.rhs
+            self.is_linear = self.is_linear or True
+        # nonlinear
+        else:
+            if eq.rhs != 0:
+                raise TypeError("RHS of a nonlinear form equation has to be 0")
+            if not isinstance(eq.lhs, (ufl.Form, slate.slate.TensorBase)):
+                raise TypeError("Provided BC residual is a '%s', not a Form or Slate Tensor" % type(eq.lhs).__name__)
+            if len(eq.lhs.arguments()) != 1:
+                raise ValueError("Provided BC residual is not a linear form")
+            self.F = eq.lhs
+            if J is not None and not isinstance(J, (ufl.Form, slate.slate.TensorBase)):
+                raise TypeError("Provided BC Jacobian is a '%s', not a Form or Slate Tensor" % type(J).__name__)
+            if J is not None and len(J.arguments()) != 2:
+                raise ValueError("Provided BC Jacobian is not a bilinear form")
+            self.J = J or ufl_expr.derivative(self.F, self.u)
+            if Jp is not None and not isinstance(Jp, (ufl.Form, slate.slate.TensorBase)):
+                raise TypeError("Provided BC preconditioner is a '%s', not a Form or Slate Tensor" % type(Jp).__name__)
+            if Jp is not None and len(Jp.arguments()) != 2:
+                raise ValueError("Provided BC preconditioner is not a bilinear form")
+            self.Jp = Jp or self.J
```

Comment:
I have a vague idea of deriving `EquationBC` and `NonlinearVariationalProblem' from the same ancestor maybe called `Equation`, as these two both have 'F`, `J`, `Jp`, and now `bcs`.  
I think these two are mathematically very similar, and I think we can make the codes more monolithic if we regard them as the same thing. For instance, we currently have recursions over EquationBCs and deal with NonlinearVariationalProblem separately, but we can probably bring these operations together.  
In that case we can place this argument checking in the ancestor class.
I need to think more about this.  What would you think?

Reply:
I was thinking in the short term just to pull that checking code out into a free function that can be used in two places. Rather than a larger-scale refactor.

I think it is worthwhile thinking about this latter change for the slightly longer term, but I would prefer to get these changes into a reasonable state sooner rather than later: I have some code that I want to base on top of this which is waiting for the merge.

Reply:
Ok! I'll do it now.